### PR TITLE
Lobby sockets

### LINKED ISSUE #24: Implement Lobby Sockets
See #20

### PR DESCRIPTION
**Pull request by taitfuller**
Lobby sockets are ✨done✨

Listens to the following client initiated events:
- connected
- disconnecting
- disconnected
- settings:update

Emits the following server initiated events:
- navigate
- host
- players:initial
- players:add
- players:remove
- settings:initial
- settings:update

For a detailed explanation of these events see [Socket.IO Events Wiki](https://github.com/SE750-DART/nerds-oppose-society/wiki/Socket.IO-Events)

Apologies for this turning out into such a behemoth - turns out a good portion of the logic implemented here will be used in other parts of the game so had to take time to get it right! Some changes had to be made to the models and additions made to the services accordingly.

Closes #24

### CODE REVIEW

File: backend/src/handlers/game.handler.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    if (isHost(socket, gameCode)) {
+      const game = await getGame(gameCode);
+      switch (setting) {
+        case "MAX_PLAYERS":
+          await setMaxPlayersService(game, value);
+          break;
+        case "ROUND_LIMIT":
+          await setRoundLimitService(game, value);
+          break;
+      }
```
Comment (Alxjlv) [inline backend/src/handlers/game.handler.ts:33]: Is this only going to update one of those two settings at a time? If it isn't, then would it be better to if/else if for when both MAX_PLAYERS and ROUND_LIMIT are changed at the same time?

File: backend/src/handlers/index.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Server, Socket } from "socket.io";
+import registerPlayerHandler, { playerJoin } from "./player.handler";
+
+export { default as Auth } from "./auth.handler";
+
+export const Connection = async (io: Server, socket: Socket): Promise<void> => {
+  registerPlayerHandler(io, socket);
+
+  await playerJoin(io, socket);
+};
```
Comment (Alxjlv) [inline backend/src/handlers/index.ts:10]: Should this be registering more handlers?

File: backend/src/models/__tests__/game.model.test.ts [hunk 3]
```diff
@@ -2,6 +2,10 @@ import mongoose from "mongoose";
 import { GameModel, SetupType } from "../../models";
 
 describe("Game Model", () => {
+  let gameCode = 42069;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  let gameData: any;
+
```
Comment (Alxjlv) [inline backend/src/models/__tests__/game.model.test.ts:8]: It looks like you did change the model slightly so these test changes could be okay. I feel like the service tests should be part of a separate PR (see my other comment on the service test)

File: backend/src/models/game.model.ts [hunk 4]
```diff
@@ -15,20 +15,22 @@ export enum GameState {
...
   rounds?: Types.DocumentArray<Round>;
 }
 
 const GameSchema: Schema = new Schema({
   gameCode: { type: String, required: true, unique: true },
-  settings: SettingsSchema,
+  settings: {
+    type: SettingsSchema,
+    default: () => ({}),
+  },
```
Comment (Alxjlv) [inline backend/src/models/game.model.ts:33]: What is this change doing? (ie. is this necessary for this PR?)

File: backend/src/models/player.model.ts [hunk 5]
```diff
@@ -3,11 +3,13 @@ import { Document, Schema } from "mongoose";
...
+  score: number;
+  new: boolean;
 }
 
 export const PlayerSchema: Schema = new Schema({
   nickname: { type: String, required: true },
   punchlines: [String],
-  score: Number,
+  score: { type: Number, default: 0 },
+  new: { type: Boolean, default: true },
```
Comment (Alxjlv) [inline backend/src/models/player.model.ts:14]: This change does look necessary 👍

File: backend/src/services/__tests__/game.service.test.ts [hunk 6]
```diff
@@ -82,14 +88,40 @@ describe("validateGameCode Service", () => {
   it("returns false for an invalid gameCode", async () => {
     const result = await validateGameCode("69420");
 
-    expect(result).toBe(false);
+    expect(result).toBeFalsy();
```
Comment (Alxjlv) [inline backend/src/services/__tests__/game.service.test.ts:91]: I feel like all these service tests should be part of a separate PR as they muddy the review and make it hard to go through all the files.

The changes I see in the services seem to be mainly refactors so I'm not sure that they need to be included in the PR. The additional tests thus seem geared toward code coverage, which I think means they should be part of a separate change.

Let me know if I've missed something however

**Review by Alxjlv — COMMENTED** [review]
Looks generally good. I would have preferred a higher number of PRs for this issue (ie. one per handler) from readability perspective (my cognitive load reading this meant I couldn't really go through in as much depth as I'd have liked).

Got a couple of comments I'd like addressed

File: backend/src/handlers/game.handler.ts (hunk 1 above)
Comment (taitfuller) [reply backend/src/handlers/game.handler.ts:33]: This will only update one setting at a time as the user can only modify one at a time 👍🏼

File: backend/src/handlers/index.ts (hunk 2 above)
Comment (taitfuller) [reply backend/src/handlers/index.ts:10]: Good catch! Fixed in 9654d6a

File: backend/src/models/__tests__/game.model.test.ts (hunk 3 above)
Comment (taitfuller) [reply backend/src/models/__tests__/game.model.test.ts:8]: Yeah I agree - it would've been nice to have split the services & handlers up although unfortunately short on time 😥

File: backend/src/models/game.model.ts (hunk 4 above)
Comment (taitfuller) [reply backend/src/models/game.model.ts:33]: This was necessary as the settings object was not set by default on a new Game

File: backend/src/models/player.model.ts (hunk 5 above)
Comment (taitfuller) [reply backend/src/models/player.model.ts:14]: Yep needed scores to default to zero & track new players that have not connected to Socket.IO yet

File: backend/src/services/__tests__/game.service.test.ts (hunk 6 above)
Comment (taitfuller) [reply backend/src/services/__tests__/game.service.test.ts:91]: Yeah I agree it would've been nicer to split these up. Unfortunately I had to make changes to the services to implement the handlers and I made some additionally refactors to the services to ensure consistency

**Review by Alxjlv — APPROVED** [review]
Solid changes man!
